File: src/engine/encode.js
// Canonical NN input encoding for the Fox-Lite self-play engine.
//
// Mover-frame ("self" = the player to move) and SUIT-CANONICALIZED: suits are
// permuted so the trump suit is always canonical slot 0, and the two non-trump
// suits map to slots 1 and 2 in ascending real-suit order. The policy head's
// 33 outputs are over canonical card slots; map back with realCardFromCanonIndex.
//
// This is the single source of truth the Rust (foxlite_core) and Python
// (training/encode.py) encoders are parity-checked against.

import { SUITS, RANKS, cardId, TRICKS_PER_ROUND, TARGET_SCORE, HUMAN } from './game.js'

export const NUM_SUITS = SUITS.length // 3
export const NUM_RANKS = RANKS.length // 11
export const NUM_CARDS = NUM_SUITS * NUM_RANKS // 33

// Layout (canonical): a history-token block followed by one-hot static blocks.
//
// History tokens (v3): one token per COMPLETED trick of the current round, in
// DESCENDING order (slot 0 = most recent completed trick). The in-progress
// trick's lead is NOT a token — it lives in the static led-card block. Each
// token is [lead card index, follow card index, led-by-self, valid]. Padded
// slots are all-zero; the valid bit disambiguates padding from a real card
// index 0 and is the net's attention/readout mask.
export const HIST_TOKENS = TRICKS_PER_ROUND - 1 // 12 (max completed tricks at a decision)
export const TOKEN_FEATS = 4 // [lead card 0..32, follow card 0..32, led-by-self 0/1, valid 0/1]
const HIST = HIST_TOKENS * TOKEN_FEATS // 48
const OWN_HAND = NUM_CARDS // 33
const TRUMP_RANK = NUM_RANKS // 11
const LED = NUM_CARDS + 1 // 34 (led card one-hot + "no led / I'm leading" slot)
const SELF_TRICKS = TRICKS_PER_ROUND + 1 // 14 (0..13)
const OPP_TRICKS = TRICKS_PER_ROUND + 1 // 14
const TRICK_NUM = TRICKS_PER_ROUND // 13 (1..13)
const SCORE_SLOTS = TARGET_SCORE // 21 (0..20)

export const INPUT_SIZE =
  HIST +
  OWN_HAND +
  TRUMP_RANK +
  LED +
  SELF_TRICKS +
  OPP_TRICKS +
  TRICK_NUM +
  SCORE_SLOTS +
  SCORE_SLOTS // 209

const suitIndex = (suit) => SUITS.indexOf(suit)

// Map a real suit index to its canonical slot given the trump suit index.
export function canonSuit(realSuitIdx, trumpIdx) {
  if (realSuitIdx === trumpIdx) return 0
  let slot = 1
  for (let s = 0; s < NUM_SUITS; s++) {
    if (s !== trumpIdx && s < realSuitIdx) slot++
  }
  return slot
}

function realSuitFromCanon(canonSlot, trumpIdx) {
  if (canonSlot === 0) return trumpIdx
  const nonTrump = []
  for (let s = 0; s < NUM_SUITS; s++) if (s !== trumpIdx) nonTrump.push(s)
  return nonTrump[canonSlot - 1]
}

// Canonical card index (0..32) for a card object {suit, rank}.
export function canonCardIndex(card, trumpIdx) {
  return canonSuit(suitIndex(card.suit), trumpIdx) * NUM_RANKS + (card.rank - 1)
}

// Inverse: canonical card slot -> a real card {suit, rank, id}.
export function realCardFromCanonIndex(ci, trumpIdx) {
  const canonSlot = Math.floor(ci / NUM_RANKS)
  const rank = (ci % NUM_RANKS) + 1
  const realSuitIdx = realSuitFromCanon(canonSlot, trumpIdx)
  const suit = SUITS[realSuitIdx]
  return { suit, rank, id: cardId(suit, rank) }
}

// Encode `state` from `mover`'s perspective. Returns Float32Array(INPUT_SIZE).
export function encode(state, mover = state.awaiting) {
  const out = new Float32Array(INPUT_SIZE)
  const trumpIdx = suitIndex(state.trump.suit)
  const moverIsHuman = mover === HUMAN

  const ownHand = moverIsHuman ? state.humanHand : state.botHand
  const selfTricks = moverIsHuman ? state.tricksWon.human : state.tricksWon.bot
  const oppTricks = moverIsHuman ? state.tricksWon.bot : state.tricksWon.human
  const selfScore = moverIsHuman ? state.score.human : state.score.bot
  const oppScore = moverIsHuman ? state.score.bot : state.score.human

  let cur = 0
  // history tokens: completed tricks, most recent first. Events arrive in play
  // order as (lead, follow) pairs; a trailing single event is the in-progress
  // trick's lead and is skipped here (it equals state.ledCard).
  const events = state.trickHistory
  const nComplete = Math.floor(events.length / 2)
  if (nComplete > HIST_TOKENS) {
    throw new Error(`completed tricks ${nComplete} > ${HIST_TOKENS}`)
  }
  for (let t = 0; t < nComplete; t++) {
    const lead = events[2 * (nComplete - 1 - t)]
    const follow = events[2 * (nComplete - 1 - t) + 1]
    const base = cur + t * TOKEN_FEATS
    out[base] = canonCardIndex(lead.card, trumpIdx)
    out[base + 1] = canonCardIndex(follow.card, trumpIdx)
    out[base + 2] = lead.player === mover ? 1 : 0
    out[base + 3] = 1
  }
  cur += HIST
  // own hand
  for (const c of ownHand) out[cur + canonCardIndex(c, trumpIdx)] = 1
  cur += OWN_HAND
  // trump rank (suit is implied = canonical slot 0)
  out[cur + (state.trump.rank - 1)] = 1
  cur += TRUMP_RANK
  // led card + "no led / I'm leading" flag
  if (state.ledCard) out[cur + canonCardIndex(state.ledCard, trumpIdx)] = 1
  else out[cur + NUM_CARDS] = 1
  cur += LED
  // tricks won
  out[cur + Math.min(selfTricks, TRICKS_PER_ROUND)] = 1
  cur += SELF_TRICKS
  out[cur + Math.min(oppTricks, TRICKS_PER_ROUND)] = 1
  cur += OPP_TRICKS
  // trick number (1..13)
  out[cur + (state.trickNum - 1)] = 1
  cur += TRICK_NUM
  // match scores (one-hot 0..20, clamped)
  out[cur + Math.min(selfScore, SCORE_SLOTS - 1)] = 1
  cur += SCORE_SLOTS
  out[cur + Math.min(oppScore, SCORE_SLOTS - 1)] = 1
  cur += SCORE_SLOTS

  if (cur !== INPUT_SIZE) throw new Error(`encode cursor ${cur} != ${INPUT_SIZE}`)
  return out
}

// Canonical legal-move mask (Float32Array(33)) for `mover`.
export function legalMask(state, mover = state.awaiting) {
  const out = new Float32Array(NUM_CARDS)
  const trumpIdx = suitIndex(state.trump.suit)
  const hand = mover === HUMAN ? state.humanHand : state.botHand
  const led = state.ledCard
  let legal = hand
  if (led) {
    const same = hand.filter((c) => c.suit === led.suit)
    if (same.length > 0) legal = same
  }
  for (const c of legal) out[canonCardIndex(c, trumpIdx)] = 1
  return out
}

File: training/encode.py
"""Canonical NN input encoding (suit-canonicalized, mover-frame).

Byte-for-byte parity with src/engine/encode.js and foxlite_core/src/encode.rs.
This Python copy is the reference used to build forward-parity fixtures and to
sanity-check the Rust/JS encoders. It consumes a state dict shaped exactly like
the JS game state (humanHand/botHand lists of {suit,rank,id}, trump, ledCard,
tricksWon, score, trickNum, trickHistory, awaiting).

v3 layout: [ history tokens | static one-hot blocks ]. One token per COMPLETED
trick of the current round, in DESCENDING order (slot 0 = most recent completed
trick); the in-progress trick's lead is NOT a token — it lives in the static
led-card block. Each token is [lead card index, follow card index, led-by-self,
valid]. Padded slots are all-zero; the valid bit disambiguates padding from a
real card index 0 and is the net's attention/readout mask.
"""

from __future__ import annotations

from typing import Optional

import numpy as np

SUITS = ("bells", "keys", "moons")
RANKS = tuple(range(1, 12))
HUMAN = "human"
BOT = "bot"
NUM_SUITS = len(SUITS)
NUM_RANKS = len(RANKS)
NUM_CARDS = NUM_SUITS * NUM_RANKS  # 33
TRICKS_PER_ROUND = 13
TARGET_SCORE = 21

# Block sizes (canonical layout).
HIST_TOKENS = TRICKS_PER_ROUND - 1  # 12 (max completed tricks at a decision)
TOKEN_FEATS = 4  # [lead card 0..32, follow card 0..32, led-by-self 0/1, valid 0/1]
HIST = HIST_TOKENS * TOKEN_FEATS  # 48
_OWN_HAND = NUM_CARDS
_TRUMP_RANK = NUM_RANKS
_LED = NUM_CARDS + 1  # led card one-hot + "no led / I'm leading" slot
_SELF_TRICKS = TRICKS_PER_ROUND + 1
_OPP_TRICKS = TRICKS_PER_ROUND + 1
_TRICK_NUM = TRICKS_PER_ROUND
_SCORE_SLOTS = TARGET_SCORE

INPUT_SIZE = (
    HIST + _OWN_HAND + _TRUMP_RANK + _LED
    + _SELF_TRICKS + _OPP_TRICKS + _TRICK_NUM + _SCORE_SLOTS + _SCORE_SLOTS
)  # 209


def _suit_index(suit: str) -> int:
    return SUITS.index(suit)


def canon_suit(real_suit_idx: int, trump_idx: int) -> int:
    if real_suit_idx == trump_idx:
        return 0
    slot = 1
    for s in range(NUM_SUITS):
        if s != trump_idx and s < real_suit_idx:
            slot += 1
    return slot


def real_suit_from_canon(canon_slot: int, trump_idx: int) -> int:
    if canon_slot == 0:
        return trump_idx
    non_trump = [s for s in range(NUM_SUITS) if s != trump_idx]
    return non_trump[canon_slot - 1]


def canon_card_index(card: dict, trump_idx: int) -> int:
    return canon_suit(_suit_index(card["suit"]), trump_idx) * NUM_RANKS + (card["rank"] - 1)


def real_card_from_canon_index(ci: int, trump_idx: int) -> dict:
    canon_slot = ci // NUM_RANKS
    rank = ci % NUM_RANKS + 1
    suit = SUITS[real_suit_from_canon(canon_slot, trump_idx)]
    return {"suit": suit, "rank": rank, "id": f"{suit}-{rank}"}


def encode(state: dict, mover: Optional[str] = None) -> np.ndarray:
    if mover is None:
        mover = state["awaiting"]
    out = np.zeros(INPUT_SIZE, dtype=np.float32)
    trump_idx = _suit_index(state["trump"]["suit"])
    mover_is_human = mover == HUMAN

    own_hand = state["humanHand"] if mover_is_human else state["botHand"]
    self_tricks = state["tricksWon"]["human" if mover_is_human else "bot"]
    opp_tricks = state["tricksWon"]["bot" if mover_is_human else "human"]
    self_score = state["score"]["human" if mover_is_human else "bot"]
    opp_score = state["score"]["bot" if mover_is_human else "human"]

    cur = 0
    # history tokens: completed tricks, most recent first. Events arrive in play
    # order as (lead, follow) pairs; a trailing single event is the in-progress
    # trick's lead and is skipped here (it equals state["ledCard"]).
    events = state["trickHistory"]
    n_complete = len(events) // 2
    if n_complete > HIST_TOKENS:
        raise RuntimeError(f"completed tricks {n_complete} > {HIST_TOKENS}")
    for t in range(n_complete):
        lead = events[2 * (n_complete - 1 - t)]
        follow = events[2 * (n_complete - 1 - t) + 1]
        base = cur + t * TOKEN_FEATS
        out[base] = canon_card_index(lead["card"], trump_idx)
        out[base + 1] = canon_card_index(follow["card"], trump_idx)
        out[base + 2] = 1.0 if lead["player"] == mover else 0.0
        out[base + 3] = 1.0
    cur += HIST
    for c in own_hand:
        out[cur + canon_card_index(c, trump_idx)] = 1.0
    cur += _OWN_HAND
    out[cur + (state["trump"]["rank"] - 1)] = 1.0
    cur += _TRUMP_RANK
    if state["ledCard"] is not None:
        out[cur + canon_card_index(state["ledCard"], trump_idx)] = 1.0
    else:
        out[cur + NUM_CARDS] = 1.0
    cur += _LED
    out[cur + min(self_tricks, TRICKS_PER_ROUND)] = 1.0
    cur += _SELF_TRICKS
    out[cur + min(opp_tricks, TRICKS_PER_ROUND)] = 1.0
    cur += _OPP_TRICKS
    out[cur + (state["trickNum"] - 1)] = 1.0
    cur += _TRICK_NUM
    out[cur + min(self_score, _SCORE_SLOTS - 1)] = 1.0
    cur += _SCORE_SLOTS
    out[cur + min(opp_score, _SCORE_SLOTS - 1)] = 1.0
    cur += _SCORE_SLOTS

    if cur != INPUT_SIZE:
        raise RuntimeError(f"encode cursor {cur} != {INPUT_SIZE}")
    return out


def legal_mask(state: dict, mover: Optional[str] = None) -> np.ndarray:
    if mover is None:
        mover = state["awaiting"]
    out = np.zeros(NUM_CARDS, dtype=np.float32)
    trump_idx = _suit_index(state["trump"]["suit"])
    hand = state["humanHand"] if mover == HUMAN else state["botHand"]
    led = state["ledCard"]
    legal = hand
    if led is not None:
        same = [c for c in hand if c["suit"] == led["suit"]]
        if same:
            legal = same
    for c in legal:
        out[canon_card_index(c, trump_idx)] = 1.0
    return out
